libraries/standard/json/include/json.h

```diff
@@ -0,0 +1,111 @@
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#ifndef JSON_H
+#define JSON_H
+
+#include <stddef.h>
+
+typedef enum JSONStatus
+{
+    JSONPartial = 0,
+    JSONSuccess,
+    JSONIllegalDocument,
+    JSONMaxDepthExceeded,
+    JSONNotFound
+} JSONStatus_t;
+
+/**
+ * @brief Parse a buffer to determine if it contains a valid JSON document.
+ *
+ * @param[in] buf  The buffer to parse.
+ * @param[in] length  The size of the buffer.
+ *
+ * @return #JSONSuccess if the buffer contents are valid JSON;
+ * #JSONIllegalDocument if the buffer contents are NOT valid JSON;
+ * #JSONMaxDepthExceeded if object and array nesting exceeds a threshold;
+ * #JSONPartial if the buffer contents are potentially valid but incomplete.
+ *
+ * @note The maximum nesting depth may be specified by defining the macro
+ * JSON_MAX_DEPTH.  The default is 32 of sizeof(char).
+ *
+ * @note By default, a valid JSON document may contain a single element
+ * (e.g., string, boolean, number).  To require that a valid document
+ * contain an object or array, define JSON_VALIDATE_COLLECTIONS_ONLY.
+ */
+JSONStatus_t JSON_Validate( const char * buf,
+                            size_t length );
+
+/**
+ * @brief Find a key in a JSON object and output a pointer to its value.
+ *
+ * @param[in] buf  The buffer to search.
+ * @param[in] length  size of the buffer.
+ * @param[in] queryKey  The key to search for.
+ * @param[in] queryKeyLength  Length of the key.
+ * @param[in] separator  A character between a key and a sub-key in queryKey.
+ * @param[out] outValue  A pointer to receive the address of the value found.
+ * @param[out] outValueLength  A pointer to receive the length of the value found.
+ *
+ * The JSON document must contain an object (e.g., '{"key":"value"}'). Any
+ * value may also be an object and so forth to a maximum depth.  A search
+ * may descend through nested objects when the queryKey contains matching
+ * key strings joined by a separator.
+ *
+ * For example, if buf contains '{"foo":"abc","bar":{"foo","xyz"}}', then a 
```

Comment:
can you also document what happens if I search the example for 'bar'? I'm guessing a pointer to "{\"foo\",\"xyz\"}}", but there's no reason I should have to guess.

Reply:
added